packages/backend/src/core/totalSupply/providers/ArbitrumTotalSupplyProvider.ts

```diff
@@ -0,0 +1,127 @@
+import { EthereumClient } from '../../../peripherals/ethereum/EthereumClient'
+import { MulticallRequest } from '../../../peripherals/ethereum/MulticallClient'
+import { TotalSupplyProvider, TotalSupplyQuery } from '../TotalSupplyProvider'
+
+const erc20Interface = new utils.Interface([
+  'function totalSupply() view returns (uint256)',
+])
+
+export class ArbitrumTotalSupplyProvider implements TotalSupplyProvider {
+  private readonly chainId = ChainId.ARBITRUM
+
+  constructor(
+    private readonly arbitrumClient: EthereumClient,
+    private readonly multicallClient: ArbitrumMulticallClient,
+  ) {
+    assert(
+      this.chainId === this.multicallClient.getChainId(),
+      'chainId mismatch between multicallClient and balance provider',
+    )
+  }
+
+  public getChainId(): ChainId {
+    return this.chainId
+  }
+
+  public async getTotalSupplies(
+    totalSupplyQueries: TotalSupplyQuery[],
+    timestamp: UnixTime,
+    blockNumber: number,
+  ): Promise<TotalSupplyRecord[]> {
+    const canMulticallBeUsed = this.multicallClient.canBeUsed(blockNumber)
+
+    if (canMulticallBeUsed) {
+      return this.fetchUsingMulticall(
+        totalSupplyQueries,
+        timestamp,
+        blockNumber,
+      )
+    }
+
+    return this.fetchUsingPlainCalls(totalSupplyQueries, timestamp, blockNumber)
+  }
+
+  private async fetchUsingMulticall(
+    totalSupplyQueries: TotalSupplyQuery[],
+    timestamp: UnixTime,
+    blockNumber: number,
+  ): Promise<TotalSupplyRecord[]> {
+    const calls = totalSupplyQueries.map(({ tokenAddress }) =>
+      encodeErc20TotalSupplyQuery(tokenAddress),
+    )
+
+    const totalSupplyResponses = await this.multicallClient.multicall(
+      calls,
+      blockNumber,
+    )
+
+    return totalSupplyResponses.map((totalSupplyResponse, i) => {
+      const totalSupply = totalSupplyResponse.success
+        ? decodeErc20TotalSupplyQuery(totalSupplyResponse.data)
+        : 0n
+
+      return {
+        timestamp,
+        totalSupply,
+        assetId: totalSupplyQueries[i].assetId,
+        chainId: this.chainId,
+      }
+    })
+  }
+
+  private async fetchUsingPlainCalls(
+    totalSupplyQueries: TotalSupplyQuery[],
+    timestamp: UnixTime,
+    blockNumber: number,
+  ): Promise<TotalSupplyRecord[]> {
+    const rpcCalls = totalSupplyQueries.map(({ tokenAddress }) => {
+      const { address: to, data } = encodeErc20TotalSupplyQuery(tokenAddress)
+
+      return this.arbitrumClient.call({ to, data }, blockNumber)
+    })
+
+    const responses = await Promise.all(rpcCalls)
+
+    return totalSupplyQueries.map(({ assetId }, i) => {
+      const totalSupplyResponse = responses[i]
+
+      const totalSupply = decodeErc20TotalSupplyQuery(totalSupplyResponse)
+
+      return {
+        assetId,
+        timestamp,
+        chainId: this.chainId,
+        totalSupply,
+      }
+    })
+  }
+}
+
+function encodeErc20TotalSupplyQuery(
```

Comment:
Add a test for this function

Reply:
https://github.com/l2beat/l2beat/pull/1664/commits/73cc39ab4c64603ed7a9d197e6dbdaabba271575